scripts/xcode/with-environment.sh

```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# This script is used to source in Xcode the environment settings required to run properly.
+# The script first sources the base `.xcode.env` file.
+# Then it sources the `.xcode.env.local` file if present, to override some local config
+# Finally, it will execute the command passed i input if any.
+#
+# USAGE:
+# ./with-environment.sh command
+
+ENV_PATH="$PODS_ROOT/../.xcode.env"
+if [ -f "$ENV_PATH" ]; then
+    source "$ENV_PATH"
+else
+    echo "[Error] Could not find the .xcode.env file. " \
+       "Make sure to have that file properly configured with a valid configuration for the NODE_BINARY variable. " \
+       "You can set it up quickly by running: " \
+       "echo 'export NODE_BINARY=$(command -v node)' > .xcode.env " \
+       "in the ios folder." >&2
+    exit 1
+fi
```

Comment:
As long as `command -v node` returns a valid path, I don't think we should fail. We should move the else-block further down.

Reply:
I'm not sure I got your suggestion. 
This error occurs when the `.xcode.env` file is missing and it contains a suggestion about how to configure it.

Plus, I removed `node` from my system folder and  to me `$(command -v node)` returns an empty path when run from Xcode, while it returns a valid `brew` path when run outside Xcode.

Reply:
I think `.xcode.env` should be opt-in. As long as `node` can be found, we shouldn't stop because `.xcode.env` is missing.

Reply:
In other words:

```sh
ENV_PATH="$PODS_ROOT/../.xcode.env"
if [ -f "$ENV_PATH" ]; then
    source "$ENV_PATH"
end

LOCAL_ENV_PATH="${LOCAL_ENV_PATH}.local"
if [ -f "$LOCAL_ENV_PATH" ]; then
    source "$LOCAL_ENV_PATH"
end

if command -v node >/dev/null 2>&1; then
  # Execute argument, if present
  if [ -n "$1" ]; then
    $@
  fi
else
  # Tell people how to set up .xcode.env so Xcode can find Node here
end

Reply:
I'm not sure I like this approach. I think that introduces complexity and ambiguity.

Imagine that I have two different node managers and I setup one of them in my `.env.xcode` while the `command -v node` returns the other one (That was actually my setup before start working on this pr: the `command -v node` returned `/usr/local/bin/node` from Xcode but outside of Xcode was returning the nvm installation in my brew).

We are continuing the execution under the assumption that Xcode will use the node returned by `command -v node` while it will actually use the one in the `.xcode.env` file. I can see users get lost because of this misalignment. Also, they will need to actually know how Xcode works behind the curtains to understand that `command -v node` could return something different within the IDE.

And we lose the possibility to fail fast in case of misconfigurations.

I think that it's better to ask everyone to have a proper `.xcode.env` file configured on their system for the simplicity it could bring and the disambiguation it provides. Plus, it is not that we are adding a strong opinion on anything: people can keep using their favourite manager.

What do you think? does it make sense?

I'd like to have also @cortinico opinion on this.

Reply:
Sorry, I think I made a mistake in the code above. What I wanted was to **fallback** to `command -v node` if both `.xcode.env` and `.xcode.env.local` are missing. Essentially, something like:

```sh
# Source `.xcode.env` and/or `.xcode.env.local` here

if [ ! -f $NODE_BINARY ]; then
  export NODE_BINARY=$(command -v node)
fi
```

I don't agree that `.xcode.env` should be required. There are plenty devs, myself included, that use Homebrew or `n` to install Node, and do not need to set up the environment.

Reply:
@tido64 if you install node via homebrew, it won't be seen by xcode as its not on PATH.

Reply:
Regarding `n` (I haven't used it personally): according to the docs it installs node by default to /usr/local/bin, which again is not in the xcode PATH.

Reply:
I'm not sure that's true, this is the `PATH` Xcode uses: `export PATH\=/Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin:/Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/local/bin:/Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/libexec:/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/iPhoneSimulator.platform/usr/bin:/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/iPhoneSimulator.platform/usr/local/bin:/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/usr/bin:/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/usr/local/bin:/Applications/Xcode_13.2.1.app/Contents/Developer/usr/bin:/Applications/Xcode_13.2.1.app/Contents/Developer/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin`

I've never had to mess around with Xcode's `PATH` before. All my issues have been with `find-node.sh` failing to find some random nvm config file that I don't use.

Reply:
@tido64 Ah sorry, you are right about `n` then. But homebrew installs to `/opt/homebrew/bin` on m1, which makes the `.xcode.env` necessary for homebrew on m1 and the oother managers like nvm, etc.

Reply:
I still maintain that `.xcode.env` shouldn't be required. I'd like to see this work out of box without additional setup if possible. If we are worried confusing people, I'd suggest:

1. Apply `.xcode.env` if it exists
2. Apply .xcode.env.local` if it exists
3. If `NODE_BINARY` is empty, use `command -v node` to populate `NODE_BINARY` as fallback and output a warning about potentially using the wrong installation of Node and how to set up `.xcode.env`
4. If `NODE_BINARY` is still empty, fail with how to set up `.xcode.env`

Reply:
I want to share a couple of different approaches:
1. I still believe that `.xcode.env` should be the preferred approach. One thing that we can do is to add a [`postinstall`](https://yarnpkg.com/advanced/lifecycle-scripts) script that `find` the `xcodeproj` for the `ios` app and creates the `.xcode.env` file there, if it is not present. This make the `.xcode.env` working out of the box, as @tido64 desired: as soon as the user updates RN, the script will run. This could have some downsides for what concerns `yarn` dependencies installation, though.
2. Alternatively, I would change the approach to (slimming the script for readability):
    ```sh
    export NODE_BINARY=$(command -v node)
    source .xcode.env
    source .xoce.env.local

    if [ -z $NODE_BINARY]; then
      echo "Error: no NODE_BINARY detected, please setup '.xcode.env' file"
      exit -1
    fi

    if [ -n $1 ]; then
      $1
    fi
    ```

    In this way, the script is more linear: it starts with a default, which in many case will be the same we are going to put in the env file, and then we progressively overwrite default env variables and local env variables.
  
What do you think? Probably the second approach is faster and a good tradeoff to start with.

Reply:
I like alternative 2. It's what I initially thought of suggesting first but discarded because of the potential confusion that you mentioned. Alternative 1 won't be possible in some setups where `postinstall` scripts are disabled or `node_modules` folder is checked in.

Reply:
The proposal of putting back `find-node-for-xcode` is more conservative and for sure gives us more space and time to improve the setup.

The `[Warning]` in the message is tricky: the script is executed in Xcode and the IDE does not have a particularly good way to present them. They got summered in a lot of other warnings and they will go unnoticed for sure...

I am not sure about what's the best path in this case.

Reply:
Just to be clear, I do not want to bring back `find-node.sh`. As I said earlier:

> All my issues have been with find-node.sh failing to find some random `nvm` config file that I don't use.

Reply:
> Just to be clear, I do not want to bring back find-node.sh

Same for me, but if removing it creates so much trouble for our users and we end up creating a blocking scenario for most of our users, I'm in for deprecating it + moving to a better approach.

Reply:
Ok, I updated the PR with another alternative. 
I'm following the approach suggested by @cortinico, putting back the `find-node-for-xcode.sh`.

1. Notice that that script is sourced as last alternative. Therefore, if `command -v node` works or the `.xcode.env` are set up properly, it won't be sourced (avoiding all the issues @tido64 and many other users may had in the past)
2. In case of weird configurations, nothing changes w.r.t. the previous setup. So these changes will be transparent to all the users
3. This gives us more leeway to smoothly transition away from the `find-node.sh` script.

I think that it is the best solution for this release. In the next one, we can finally dump the `find-node.sh` for good.

What do you think?